diff --git a/src/Fl_Browser_.cxx b/src/Fl_Browser_.cxx
index 56b751415..0e49fe2fb 100644
--- a/src/Fl_Browser_.cxx
+++ b/src/Fl_Browser_.cxx
@@ -707,22 +707,42 @@ int Fl_Browser_::handle(int event) {
   // must do shortcuts first or the scrollbar will get them...
   if (event == FL_ENTER || event == FL_LEAVE) return 1;
   if (event == FL_KEYBOARD && type() >= FL_HOLD_BROWSER) {
-    void* l1 = selection_;
-    void* l = l1; if (!l) l = top_; if (!l) l = item_first();
+    void* l = selection_;
+    if (!l)
+      l = top_;
+    if (!l)
+      l = item_first();
     if (l) {
       if (type()==FL_HOLD_BROWSER) {
         switch (Fl::event_key()) {
         case FL_Down:
-          while ((l = item_next(l))) {
-            if (item_height(l)>0) {select_only(l, when()); break;}
+          if (selection_)
+            l = item_next(selection_);
+          while (l) {
+            if (item_height(l)>0) {
+              select_only(l, when());
+              break;
+            }
+            l = item_next(l);
           }
             return 1;
         case FL_Up:
-          while ((l = item_prev(l))) {
+          if (selection_)
+            // find the item before the selected one
+            l = item_prev(selection_);
+          else {
+            // find the last visible item
+            l = find_item(y()+h()-incr_height()/2);
+            if (!l)
+              // if that fails, start from the end of the list
+              l = item_last();
+          }
+          while (l) {
             if (item_height(l)>0) {
               select_only(l, when());
               break; // no need to test wp (return 1)
             }
+            l = item_prev(l);
           }
           return 1;
         }
@@ -744,7 +764,7 @@ int Fl_Browser_::handle(int event) {
         case FL_Down:
           while ((l = item_next(l))) {
             if (Fl::event_state(FL_SHIFT|FL_CTRL))
-              select(l, l1 ? item_selected(l1) : 1, when());
+              select(l, selection_ ? item_selected(selection_) : 1, when());
             if (wp.deleted()) return 1;
             if (item_height(l)>0) goto J1;
           }
@@ -752,7 +772,7 @@ int Fl_Browser_::handle(int event) {
         case FL_Up:
           while ((l = item_prev(l))) {
             if (Fl::event_state(FL_SHIFT|FL_CTRL))
-              select(l, l1 ? item_selected(l1) : 1, when());
+              select(l, selection_ ? item_selected(selection_) : 1, when());
             if (wp.deleted()) return 1;
             if (item_height(l)>0) goto J1;
           }
diff --git a/src/Fl_Menu.cxx b/src/Fl_Menu.cxx
index f506c96db..d77d30593 100644
--- a/src/Fl_Menu.cxx
+++ b/src/Fl_Menu.cxx
@@ -1075,6 +1075,10 @@ const Fl_Menu_Item* Fl_Menu_Item::pulldown(
     pp.p[pp.menu_number]->autoscroll(pp.item_number);
 
   STARTUP:
+    if (pp.menu_number < 0 || pp.menu_number >= pp.nummenus) {
+      initial_item = 0; // turn off startup code
+      continue;
+    }
     menuwindow& cw = *pp.p[pp.menu_number];
     const Fl_Menu_Item* m = pp.current_item;
     if (!m || !m->activevisible()) { // pointing at inactive item
diff --git a/src/Fl_Window.cxx b/src/Fl_Window.cxx
index 593622071..79b7c9883 100644
--- a/src/Fl_Window.cxx
+++ b/src/Fl_Window.cxx
@@ -58,8 +58,8 @@ void Fl_Window::_Fl_Window() {
   minw_ = maxw_ = minh_ = maxh_ = 0;
   no_fullscreen_x = 0;
   no_fullscreen_y = 0;
-  no_fullscreen_w = w();
-  no_fullscreen_h = h();
+  no_fullscreen_w = 0;
+  no_fullscreen_h = 0;
   fullscreen_screen_top = -1;
   fullscreen_screen_bottom = -1;
   fullscreen_screen_left = -1;
diff --git a/src/Fl_cocoa.mm b/src/Fl_cocoa.mm
index 725d2c201..5dab9e44e 100644
--- a/src/Fl_cocoa.mm
+++ b/src/Fl_cocoa.mm
@@ -165,6 +165,7 @@ const NSUInteger NSEventModifierFlagOption = NSAlternateKeyMask;
 const NSUInteger NSEventModifierFlagControl = NSControlKeyMask;
 const NSUInteger NSEventModifierFlagShift = NSShiftKeyMask;
 const NSUInteger NSEventModifierFlagCapsLock = NSAlphaShiftKeyMask;
+const NSUInteger NSEventModifierFlagFunction = NSFunctionKeyMask;
 
 const NSEventType NSEventTypeLeftMouseDown = NSLeftMouseDown;
 const NSEventType NSEventTypeRightMouseDown = NSRightMouseDown;
@@ -2639,6 +2640,9 @@ static FLTextInputContext* fltextinputcontext_instance = nil;
 }
 - (void)keyDown:(NSEvent *)theEvent {
   //NSLog(@"keyDown:%@",[theEvent characters]);
+  if (![[theEvent window] isKindOfClass:[FLWindow class]]) { // occurs with #1431
+    return;
+  }
   fl_lock_function();
   Fl_Window *window = [(FLWindow*)[theEvent window] getFl_Window];
   Fl::first_window(window);
diff --git a/src/Fl_win32.cxx b/src/Fl_win32.cxx
index 3981b8966..f0b782424 100644
--- a/src/Fl_win32.cxx
+++ b/src/Fl_win32.cxx
@@ -562,7 +562,7 @@ void Fl_WinAPI_Screen_Driver::desktop_scale_factor() {
     dpi[ns][0] = float(dpiX);
     dpi[ns][1] = float(dpiY);
     scale(ns, dpiX / 96.f);
-    // fprintf(LOG, "desktop_scale_factor ns=%d factor=%.2f dpi=%.1f\n", ns, scale(ns), dpi[ns][0]);
+    // fprintf(stderr, "desktop_scale_factor ns=%d factor=%.2f dpi=%.1f\n", ns, scale(ns), dpi[ns][0]);
   }
   update_scaling_capability();
 }
@@ -1222,6 +1222,7 @@ static BOOL CALLBACK child_window_cb(HWND child_xid, LPARAM data) {
   return TRUE;
 }
 
+static bool sizing_window = false;
 
 static LRESULT CALLBACK WndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam) {
 
@@ -1244,14 +1245,14 @@ static LRESULT CALLBACK WndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lPar
     switch (uMsg) {
 
       case WM_DPICHANGED: { // 0x02E0, after display re-scaling and followed by WM_DISPLAYCHANGE
-        if (is_dpi_aware) {
+        if (is_dpi_aware && !window->parent()) {
           RECT r, *lParam_rect = (RECT*)lParam;
           Fl_WinAPI_Screen_Driver *sd = (Fl_WinAPI_Screen_Driver*)Fl::screen_driver();
           int centerX = (lParam_rect->left + lParam_rect->right)/2;
           int centerY = (lParam_rect->top + lParam_rect->bottom)/2;
           int ns = sd->screen_num_unscaled(centerX, centerY);
-          int old_ns = Fl_Window_Driver::driver(window)->screen_num();
-          if (sd->dpi[ns][0] != HIWORD(wParam) && ns == old_ns) { // change DPI of a screen
+          //fprintf(stderr,"WM_DPICHANGED ns=%d win_scale=%g\n", ns,HIWORD(wParam)/96.);
+          if (sd->dpi[ns][0] != HIWORD(wParam)) { // change DPI of a screen
             sd->dpi[ns][0] = sd->dpi[ns][1] = HIWORD(wParam);
             float f = HIWORD(wParam) / 96.f;
             GetClientRect(hWnd, &r);
@@ -1259,21 +1260,51 @@ static LRESULT CALLBACK WndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lPar
             Fl::screen_driver()->scale(ns, f);
             Fl_Window_Driver::driver(window)->resize_after_scale_change(ns, old_f, f);
             sd->update_scaling_capability();
-          } else if (ns != old_ns) {
-            // jump window with Windows+Shift+L|R-arrow to other screen with other DPI
-            if (ns >= 0) Fl_Window_Driver::driver(window)->screen_num(ns);
-            UINT flags = SWP_NOSENDCHANGING | SWP_NOZORDER | SWP_NOACTIVATE | SWP_NOOWNERZORDER | SWP_NOCOPYBITS;
-            SetWindowPos(hWnd, NULL, lParam_rect->left, lParam_rect->top,
-                         lParam_rect->right - lParam_rect->left,
-                         lParam_rect->bottom - lParam_rect->top, flags);
-            if (ns >= 0) {
-              if (Fl::modal() && Fl::modal()->menu_window()) { // is a menu window active?
-                Fl::e_x_root = 1000000;
-                Fl::modal()->handle(FL_PUSH); // simulate a distant click to close menus
+          } else if (!sizing_window) {
+            // Move window to screen with other DPI with mouse or with Windows+Shift+L|R-arrow
+            Fl_WinAPI_Window_Driver *wd = (Fl_WinAPI_Window_Driver*)Fl_Window_Driver::driver(window);
+            int olds = window->screen_num();
+            wd->screen_num(ns); // Assign window its new new screen (ns)
+            // Compute bt, bx, by for screen ns
+            int bt, bx, by;
+            scale = Fl::screen_scale(ns);
+            if (window->border()) {
+              int X, Y, W, H, dummy_x, dummy_y;
+              sd->screen_xywh_unscaled(X, Y, W, H, ns); // Screen coordinates of screen ns
+              wd->x(X + W/2 - window->w()*scale/2); // Transiently locate win at center of screen and
+              wd->y(Y + H/2 - window->h()*scale/2); // use fake_X_WM to compute size of titlebar and borders
+              wd->fake_X_wm(dummy_x, dummy_y, bt, bx, by);
+            } else { // handle borderless window moved to new screen
+              bx = by = bt = 0;
+              float old_scale = Fl::screen_scale(olds); // scale factor before change of screen
+              if (scale != old_scale) {
+                POINT curs_pos;
+                GetCursorPos(&curs_pos); // cursor position in desktop before change of screen
+                GetWindowRect(hWnd, &r); // window position in desktop before change of screen
+                if (curs_pos.x >= r.left && curs_pos.x < r.right &&
+                    curs_pos.y >= r.top && curs_pos.y < r.bottom) { // make sure cursor is inside window
+                  // fractional horizontal and vertical position of cursor in window before change of screen
+                  float fcx = (curs_pos.x - window->x() * old_scale) / (window->w() * old_scale);
+                  float fcy = (curs_pos.y - window->y() * old_scale) / (window->h() * old_scale);
+                  // set top-left of window after change of screen to maintain relative position
+                  // of cursor within window
+                  lParam_rect->left = curs_pos.x - fcx * window->w() * scale;
+                  lParam_rect->top = curs_pos.y - fcy * window->h() * scale;
+                }
               }
-              scale = Fl::screen_driver()->scale(ns);
-              EnumChildWindows(hWnd, child_window_cb, (LPARAM)&scale);
             }
+            /*fprintf(stderr, "WM_DPICHANGED calls SetWindowPos %dx%d %gx%g\n", lParam_rect->left,
+             lParam_rect->top, window->w() * scale + 2 * bx, window->h() * scale + bt + 2 * by);*/
+            wd->is_a_rescale(1);
+            window->position((lParam_rect->left + bx) / scale, (lParam_rect->top + by + bt) / scale);
+            wd->is_a_rescale(0);
+            if (Fl::modal() && Fl::modal()->menu_window()) { // is a menu window active?
+              Fl::e_x_root = 1000000;
+              Fl::modal()->handle(FL_PUSH); // simulate a distant click to close menus
+            }
+          } else { // when resizing a window with the mouse
+            Fl_Window_Driver::driver(window)->screen_num(ns); // Assign window its new new screen (ns)
+            window->redraw();
           }
         }
         return 0;
@@ -1717,12 +1748,28 @@ content  key    keyboard layout
         Fl_WinAPI_Window_Driver::driver(window)->set_minmax((LPMINMAXINFO)lParam);
         break;
 
+      case WM_SIZING: { // sent to toplevel windows when resized with the mouse
+        sizing_window = true;
+        RECT *r = (RECT*)lParam;
+        int ctrX = (r->left + r->right)/2, ctrY = (r->top + r->bottom)/2;
+        Fl_WinAPI_Screen_Driver *sd = (Fl_WinAPI_Screen_Driver *)Fl::screen_driver();
+        int ns = sd->screen_num_unscaled(ctrX, ctrY); // screen number after size change
+        int olds = window->screen_num(); // screen number before size change
+        if (ns != olds && sd->dpi[ns][0] == sd->dpi[olds][0]) { // When both screens have same DPI,
+          Fl_Window_Driver::driver(window)->screen_num(ns);     // assign window to new screen here.
+        }
+        return 1;
+      }
+
       case WM_SIZE:
         if (!window->parent()) {
           Fl_Window_Driver::driver(window)->is_maximized(wParam == SIZE_MAXIMIZED);
           if (wParam == SIZE_MINIMIZED || wParam == SIZE_MAXHIDE) {
             Fl::handle(FL_HIDE, window);
           } else {
+            /*fprintf(stderr,"WM_SIZE %dx%d moving_window=%d s=%g ns=%d\n",
+                    int(ceil(LOWORD(lParam) / scale)), int(ceil(HIWORD(lParam) / scale)),
+                    moving_window, scale, window->screen_num());*/
             if (!moving_window) {
               Fl::handle(FL_SHOW, window);
               resize_bug_fix = window;
@@ -1734,14 +1781,20 @@ content  key    keyboard layout
             }
           }
         }
-        break;
+        sizing_window = false;
+        return 0;
 
-      case WM_MOVING:
+      case WM_MOVING: // sent only to bordered toplevel windows
+        //fprintf(stderr,"WM_MOVING\n");
         moving_window = true;
         return 1;
 
       case WM_CAPTURECHANGED:
-        moving_window = false;
+        if (window->parent()) break;
+        //fprintf(stderr,"WM_CAPTURECHANGED moving_window=%d\n",moving_window);
+        if (moving_window) {
+          moving_window = false;
+        }
         resize_bug_fix = 0;
         return 0;
 
@@ -1751,10 +1804,9 @@ content  key    keyboard layout
         }
         if (moving_window) resize_bug_fix = window;
         POINTS pts = MAKEPOINTS(lParam);
-        int nx = pts.x, ny = pts.y;
-        // detect when window centre changes screen
+        Fl_Window_Driver *wd = Fl_Window_Driver::driver(window);
         Fl_WinAPI_Screen_Driver *sd = (Fl_WinAPI_Screen_Driver *)Fl::screen_driver();
-        Fl_WinAPI_Window_Driver *wd = Fl_WinAPI_Window_Driver::driver(window);
+        // detect when window centre changes screen
         int olds = wd->screen_num();
         // Issue #1097: when a fullscreen window is restored to its size, it receives first a WM_MOVE
         // and then a WM_SIZE, so it still has its fullscreen size at the WM_MOVE event, which defeats
@@ -1765,13 +1817,24 @@ content  key    keyboard layout
         if (!window->fullscreen_active() && *wd->no_fullscreen_w()) {
           trueW = *wd->no_fullscreen_w(); trueH = *wd->no_fullscreen_h();
         }
-        int news = sd->screen_num_unscaled(nx + int(trueW * scale / 2), ny + int(trueH * scale / 2));
-        if (news == -1)
+        int cX = pts.x + int(trueW * scale / 2), cY = pts.y + int(trueH * scale / 2);
+        int news = sd->screen_num_unscaled(cX, cY);
+        if (news == -1) {
           news = olds;
-        else if (news != olds) wd->screen_num(news);
-        scale = sd->scale(news);
-        wd->x(int(round(nx/scale)));
-        wd->y(int(round(ny/scale)));
+        } else if (news != olds) {
+          // Windows documentation of WM_DPICHANGED states:
+          // "The current DPI for a window always equals the last DPI sent by WM_DPICHANGED."
+          // Therefore processing of WM_MOVE should not change the window screen to one with
+          // a distinct DPI. A WM_DPICHANGED message will arrive and will change the screen.
+          // But, if both screens have equal DPI, the window should be assigned to the new screen.
+          if (sd->dpi[news][0] == sd->dpi[olds][0]) { // see #1371
+            wd->screen_num(news);
+            scale = sd->scale(news);
+            window->redraw();
+          }
+        }
+        wd->x(int(round(pts.x / scale)));
+        wd->y(int(round(pts.y / scale)));
         }
         return 0;
 
@@ -1807,7 +1870,10 @@ content  key    keyboard layout
         return 1;
 
       case WM_DISPLAYCHANGE: {// when screen configuration (number, size, position) changes
+        Fl_WinAPI_Screen_Driver *sd = (Fl_WinAPI_Screen_Driver*)Fl::screen_driver();
+        sd->screen_count_set(-1);
         Fl::call_screen_init();
+        sd->desktop_scale_factor();
         Fl::handle(FL_SCREEN_CONFIGURATION_CHANGED, NULL);
         return 0;
       }
@@ -1965,7 +2031,7 @@ int Fl_WinAPI_Window_Driver::fake_X_wm(int &X, int &Y, int &bt, int &bx, int &by
   // Find screen that contains most of the window
   // FIXME: this ought to be the "work area" instead of the entire screen !
   int scr_x = 0, scr_y = 0, scr_w = 0, scr_h = 0;
-  int ns = Fl::screen_num(int(round(X / s)), int(round(Y / s)), int(W / s), int(H / s));
+  int ns = screen_num();
   ((Fl_WinAPI_Screen_Driver*)Fl::screen_driver())->screen_xywh_unscaled(scr_x, scr_y, scr_w, scr_h, ns);
   // Make border's lower right corner visible
   if (scr_x + scr_w < X + W)
@@ -2197,13 +2263,12 @@ void Fl_WinAPI_Window_Driver::makeWindow() {
   int nscreen = 0;
   if (w->parent()) {
     nscreen = Fl_Window_Driver::driver(w->top_window())->screen_num();
-  } else if (Fl_Window_Driver::driver(w)->force_position() && Fl_WinAPI_Window_Driver::driver(w)->screen_num_ >= 0) {
-    nscreen = Fl_Window_Driver::driver(w)->screen_num();
+  } else if (Fl_Window_Driver::driver(w)->force_position()) {
+    if (Fl_WinAPI_Window_Driver::driver(w)->screen_num_ >= 0)
+      nscreen = Fl_WinAPI_Window_Driver::driver(w)->screen_num_;
+    else nscreen = Fl::screen_num(w->x(), w->y());
   } else {
-    Fl_Window *hint = Fl::first_window();
-    if (hint) {
-      nscreen = Fl_Window_Driver::driver(hint->top_window())->screen_num();
-    } else if (Fl::screen_driver()->screen_count() > 1 ) {
+    if (Fl::screen_driver()->screen_count() > 1 ) {
       // put the new window on same screen as mouse
        int mx, my, X, Y, W, H;
        nscreen = Fl::screen_driver()->get_mouse(mx, my);
@@ -2920,14 +2985,14 @@ void Fl_WinAPI_Window_Driver::capture_titlebar_and_borders(Fl_RGB_Image *&top, F
   int offset = r.left < 0 ? -r.left : 0;
   Fl_WinAPI_Screen_Driver *dr = (Fl_WinAPI_Screen_Driver *)Fl::screen_driver();
   if (htop && r.right - r.left > offset) {
-    top = dr->read_win_rectangle_unscaled(r.left+offset, r.top, r.right - r.left-offset, htop, 0);
+    top = dr->read_win_rectangle_unscaled(r.left+offset, r.top, r.right - r.left-offset, htop, 0, true);
     if (scaling != 1 && top)
       top->scale(ww, int(htop / scaling), 0, 1);
   }
   if (wsides) {
-    left = dr->read_win_rectangle_unscaled(r.left + offset, r.top + htop, wsides, int(h() * scaling), 0);
-    right = dr->read_win_rectangle_unscaled(r.right - wsides, r.top + htop, wsides, int(h() * scaling), 0);
-    bottom = dr->read_win_rectangle_unscaled(r.left+offset, r.bottom - hbottom, ww, hbottom, 0);
+    left = dr->read_win_rectangle_unscaled(r.left + offset, r.top + htop, wsides, int(h() * scaling), 0, true);
+    right = dr->read_win_rectangle_unscaled(r.right - wsides, r.top + htop, wsides, int(h() * scaling), 0, true);
+    bottom = dr->read_win_rectangle_unscaled(r.left+offset, r.bottom - hbottom, ww, hbottom, 0, true);
     if (scaling != 1) {
       if (left) left->scale(wsides, h(), 0, 1);
       if (right) right->scale(wsides, h(), 0, 1);
diff --git a/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.H b/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.H
index b038fc5f1..8fb7f759f 100644
--- a/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.H
+++ b/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.H
@@ -43,6 +43,7 @@ public:
   float dpi[MAX_SCREENS][2];
   enum APP_SCALING_CAPABILITY scaling_capability;
   void update_scaling_capability();
+  void screen_count_set(int count) { num_screens = count; }
   Fl_WinAPI_Screen_Driver();
   // --- display management
   int visual(int flags) FL_OVERRIDE;
@@ -67,7 +68,7 @@ public:
   int dnd(int unused) FL_OVERRIDE;
   int compose(int &del) FL_OVERRIDE;
   Fl_RGB_Image *read_win_rectangle(int X, int Y, int w, int h, Fl_Window *win, bool may_capture_subwins, bool *did_capture_subwins) FL_OVERRIDE;
-  Fl_RGB_Image *read_win_rectangle_unscaled(int X, int Y, int w, int h, Fl_Window *win);
+  Fl_RGB_Image *read_win_rectangle_unscaled(int X, int Y, int w, int h, Fl_Window *win, bool screen_capture);
   int get_mouse(int &x, int &y) FL_OVERRIDE;
   void enable_im() FL_OVERRIDE;
   void disable_im() FL_OVERRIDE;
diff --git a/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.cxx b/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.cxx
index a3a7ccf47..91c5bef72 100644
--- a/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.cxx
+++ b/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.cxx
@@ -112,6 +112,7 @@ void Fl_WinAPI_Screen_Driver::init()
   // we do a run-time check for the required functions...
   HMODULE hMod = GetModuleHandle("USER32.DLL");
 
+  if (num_screens >= 0) return;
   if (hMod) {
     // check that EnumDisplayMonitors is available
     fl_edm_func fl_edm = (fl_edm_func)GetProcAddress(hMod, "EnumDisplayMonitors");
@@ -126,6 +127,9 @@ void Fl_WinAPI_Screen_Driver::init()
         //      NOTE: num_screens is incremented in screen_cb so we must first reset it here...
         num_screens = 0;
         fl_edm(0, 0, screen_cb, (LPARAM)this);
+        /*fprintf(stderr,"Screen coordinates as seen by Windows: num_screens=%d\n",num_screens);
+        for(int i=0; i<num_screens;i++)
+          fprintf(stderr,"screen_cb: n=%d left=%d right=%d top=%d bottom=%d\n",i,screens[i].left, screens[i].right,screens[i].top, screens[i].bottom);*/
         return;
       }
     }
@@ -369,11 +373,15 @@ Fl_WinAPI_Screen_Driver::read_win_rectangle(
     if (ws < 1) ws = 1;
     if (hs < 1) hs = 1;
   }
-  return read_win_rectangle_unscaled(Fl_Scalable_Graphics_Driver::floor(X, s), Fl_Scalable_Graphics_Driver::floor(Y, s), ws, hs, win);
+  return read_win_rectangle_unscaled(Fl_Scalable_Graphics_Driver::floor(X, s), Fl_Scalable_Graphics_Driver::floor(Y, s), ws, hs, win, false);
 }
 
-Fl_RGB_Image *Fl_WinAPI_Screen_Driver::read_win_rectangle_unscaled(int X, int Y, int w, int h, Fl_Window *win)
+Fl_RGB_Image *Fl_WinAPI_Screen_Driver::read_win_rectangle_unscaled(int X, int Y, int w, int h,
+                                                                   Fl_Window *win, bool screen_capture)
 {
+  // When screen_capture is true, win is NULL and we read from the screen independently from any window.
+  // Otherwise win can be a window inside which to read or NULL to read from current offscreen.
+
   // Depth of image is always 3 here
 
   // Grab all of the pixels in the image...
@@ -387,15 +395,17 @@ Fl_RGB_Image *Fl_WinAPI_Screen_Driver::read_win_rectangle_unscaled(int X, int Y,
   int shift_x = 0; // X target shift if X modified
   int shift_y = 0; // Y target shift if X modified
 
-  if (X < 0) {
-    shift_x = -X;
-    w += X;
-    X = 0;
-  }
-  if (Y < 0) {
-    shift_y = -Y;
-    h += Y;
-    Y = 0;
+  if (!screen_capture) {
+    if (X < 0) {
+      shift_x = -X;
+      w += X;
+      X = 0;
+    }
+    if (Y < 0) {
+      shift_y = -Y;
+      h += Y;
+      Y = 0;
+    }
   }
 
   if (h < 1 || w < 1) return 0;            // nothing to copy
